### PR TITLE
Completes DnsRpc interface by adding methods and doc. Implements DefaultDnsRpc.

### LINKED ISSUE #594: Add methods to DnsServiceRpc

### PR DESCRIPTION
**Pull request by mderka**
Closes #594.

### CODE REVIEW

**Comment by mderka** [comment]
DefaultDnsRpc is to be tested within integration tests in #617.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+      throw translate(ex);
+    }
+  }
+
+  @Override
+  public ManagedZone getZone(String zoneName, Map<Option, ?> options) throws DnsException {
+    // just fields option
+    try {
+      return dns.managedZones().get(this.options.projectId(), zoneName)
+          .setFields(FIELDS.getString(options)).execute();
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: I would put `.execute()` on a new line, to have builder-like wrapping

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    } catch (IOException ex) {
+      throw translate(ex);
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: Should't we handle the not found case and return `null?
Something like:
```java
} catch (IOException ex) {
  DnsException serviceException = translate(ex);
  if (serviceException.code() == HTTP_NOT_FOUND) {
    return null;
  }
  throw serviceException;
}
```

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    }
+  }
+
+  @Override
+  public Project getProject(Map<Option, ?> options) throws DnsException {
+    try {
+      return dns.projects().get(this.options.projectId())
+          .setFields(FIELDS.getString(options)).execute();
+    } catch (IOException ex) {
+      throw translate(ex);
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: Should we return `null` for not found or throw exception in this case?
I would return `null` in other get calls (for consistency with other modules) but here a NOT FOUND project seems to me a more noteworthy error.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+  @Override
+  public Change getChangeRequest(String zoneName, String changeRequestId, Map<Option, ?> options)
+      throws DnsException {
+    try {
+      return dns.changes().get(this.options.projectId(), zoneName, changeRequestId)
+          .setFields(FIELDS.getString(options))
+          .execute();
+    } catch (IOException ex) {
+      throw translate(ex);
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: Should't we handle the not found case and return `null?
Something like:
```java
} catch (IOException ex) {
  DnsException serviceException = translate(ex);
  if (serviceException.code() == HTTP_NOT_FOUND) {
    return null;
  }
  throw serviceException;
}
```

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+      throws DnsException {
+    // options are fields, page token, page size, sort order
+    try {
+      Dns.Changes.List request = dns.changes().list(this.options.projectId(), zoneName)
+          .setFields(FIELDS.getString(options))
+          .setMaxResults(PAGE_SIZE.getInt(options))
+          .setPageToken(PAGE_TOKEN.getString(options));
+      if (SORTING_ORDER.getString(options) != null) {
+        // this needs to be checked and changed if more sorting options are implemented, issue #604
+        String key = "changeSequence";
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: maybe make this a `static final` constant?

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  public Tuple<String, Iterable<Change>> listChangeRequests(String zoneName, Map<Option, ?> options)
+      throws DnsException {
+    // options are fields, page token, page size, sort order
+    try {
+      Dns.Changes.List request = dns.changes().list(this.options.projectId(), zoneName)
+          .setFields(FIELDS.getString(options))
+          .setMaxResults(PAGE_SIZE.getInt(options))
+          .setPageToken(PAGE_TOKEN.getString(options));
+      if (SORTING_ORDER.getString(options) != null) {
+        // this needs to be checked and changed if more sorting options are implemented, issue #604
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: what about adding a `todo`?

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 7]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+    public Y y() {
+      return y;
+    }
+  }
+
+  /**
+   * Creates a new zone.
+   *
+   * @param zone a zone to be created
+   * @return Updated {@link ManagedZone} object
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: Do `@link` to external libraries actually work?

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 8]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+   * @throws DnsException upon failure
+   */
+  ManagedZone create(ManagedZone zone) throws DnsException;
+
+  /**
+   * Retrieves and returns an existing zone.
+   *
+   * @param zoneName name of the zone to be returned
+   * @param options a map of options for the service call
+   * @return a zone or {@code null} if not found
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: This is not the case, see my comment on implementation

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 9]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+   * @throws DnsException upon failure
+   */
+  boolean deleteZone(String zoneName) throws DnsException;
+
+  /**
+   * Lists DNS records for a given zone.
+   *
+   * @param zoneName name of the zone to be listed
+   * @param options a map of options for the service call
+   * @throws DnsException upon failure or if zone not found
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: "zone was not found"

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 10]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+      Map<Option, ?> options) throws DnsException;
+
+  /**
+   * Returns information about the current project.
+   * 
+   * @param options a map of options for the service call
+   * @return up-to-date project information
+   * @throws DnsException upon failure
+   */
+  Project getProject(Map<Option, ?> options) throws DnsException;
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: What happens if not found? See my comment above on implementation

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 11]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+
+  /**
+   * Applies change request to a zone. 
+   *
+   * @param zoneName the name of a zone to which the {@link Change} should be applied
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: See my comment above on `{@link}` for external libraries

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 12]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+   * @param changeRequest change to be applied
+   * @param options a map of options for the service call
+   * @return updated change object with server-assigned ID
+   * @throws DnsException upon failure or if zone not found
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: "zone was not found"

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 13]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+   */
+  Change applyChangeRequest(String zoneName, Change changeRequest, Map<Option, ?> options) 
+      throws DnsException;
+
+  /**
+   * Returns an existing change request.
+   *
+   * @param zoneName the name of a zone to which the {@link Change} was be applied
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: See my comment above on `{@link}` for external libraries

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 14]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+   * @param changeRequestId the unique id assigned to the change by the server
+   * @param options a map of options for the service call
+   * @return up-to-date change object
+   * @throws DnsException upon failure or if zone not found
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: "zone was not found"

If we do (see my comment on implementation) we should document that we return null if the change was not found

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 15]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+   */
+  Change getChangeRequest(String zoneName, String changeRequestId, Map<Option, ?> options) 
+      throws DnsException;
+
+  /**
+   * List an existing change requests for a zone.
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: remove "an"

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 16]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+   *
+   * @param zoneName the name of a zone to which the {@link Change}s were be applied
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: See my comment above on `{@link}` for external libraries

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 17]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
...
+   * @param options a map of options for the service call
+   * @throws DnsException upon failure or if zone not found
```
Comment (mziccard) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: "zone was not found"

**Comment by mziccard** [comment]
Not related to this PR. I think that the method [Dns.getProjectInfo()](https://github.com/GoogleCloudPlatform/gcloud-java/blob/dns-alpha/gcloud-java-dns/src/main/java/com/google/gcloud/dns/Dns.java#L524) should be called `getProject`.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java (hunk 1 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: Ok.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java (hunk 2 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: My bad, forgotten. Fixing everywhere.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java (hunk 4 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: fixed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java (hunk 6 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: added

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java (hunk 5 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: fixed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 7 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: Removed.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 8 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: fixed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 9 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: fixed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 11 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: removed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 12 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: fixed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 13 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: fixed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 14 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: fixed and updated doc to include null

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 15 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: removed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 16 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: removed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 17 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: fixed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 10 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: I think this should never happen in the first place. However, this is kind of the scenario when the parent container does not exist, is it not? Furthermore, it feels consistent with other parts of the library. Say that you want to fetch a bucket from non-existent project using Storage, you will get an exception as well, right?

The service call itself gives us 400 error.

**Comment by mderka** [comment]
Fixed all except getProject()---see above. 

Regarding `Dns.getProjectInfo()`, I would agree if `Project` was functional, but it really is just a container for project quota. There are no methods that you can call on it which would contact the service. Would that not be a bit confusing?

**Comment by mderka** [comment]
Ok. Please, check this despite the coverall drop. See the top comment regarding tests.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java (hunk 3 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: (I originally typed this below) TL;DR: I would do an exception.

I think this should never happen in the first place. However, this is kind of the scenario when the parent container does not exist, is it not? Furthermore, it feels consistent with other parts of the library. Say that you want to fetch a bucket from non-existent project using Storage, you will get an exception as well, right?

The service call itself gives us 400 error.

For now, I at least specified it in the doc.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 18]
```diff
@@ -16,6 +16,12 @@
 
 package com.google.gcloud.spi;
 
+import com.google.api.services.dns.model.Change;
+import com.google.api.services.dns.model.ManagedZone;
+import com.google.api.services.dns.model.Project;
+import com.google.api.services.dns.model.ResourceRecordSet;
+import com.google.gcloud.dns.DnsException;
+
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: I think the following RPC string value need to change:
"maxSize" -> "maxResults"
"dnsName" -> "name"

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java [hunk 19]
```diff
@@ -52,5 +58,114 @@ Integer getInt(Map<Option, ?> options) {
     }
   }
 
-  //TODO(mderka) add supported operations. Created issue #594.
+  class Tuple<X, Y> {
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: Per your issue #615, I think in this case API can look better if you have a specific 
`ListResult<T>` class that holds Iterable<T> for values and String cursor as it seems the Tuple
is used only for listing and in a consistent way.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * A default implementation of the DnsRpc interface.
+ */
+public class DefaultDnsRpc implements DnsRpc {
+
+  private static final String SORTING_KEY = "changeSequence";
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: Nit, at this layer I think SORT_BY would be a better name.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+    HttpRequestInitializer initializer = options.httpRequestInitializer();
+    this.dns = new Dns.Builder(transport, new JacksonFactory(), initializer)
+        .setRootUrl(options.host())
+        .setApplicationName(options.applicationName())
+        .build();
+    this.options = options;
+  }
+
+  @Override
+  public ManagedZone create(ManagedZone zone) throws DnsException {
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: I now wonder why we didn't provide options also for create (in the API and RPC layer)...

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+      throws DnsException {
+    try {
+      return dns.changes().get(this.options.projectId(), zoneName, changeRequestId)
+          .setFields(FIELDS.getString(options))
+          .execute();
+    } catch (IOException ex) {
+      DnsException serviceException = translate(ex);
+      if (serviceException.code() == HTTP_NOT_FOUND) {
+        // message format "The 'parameters.changeId' resource named '4' does not exist."
+        if (serviceException.getMessage().contains("changeId")) {
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: Re error-codes, can we add that to the list. There should be a better way to detect that (e.g. different reason).

**Comment by mziccard** [comment]
> Regarding Dns.getProjectInfo(), I would agree if Project was functional, but it really is just a container for project quota. There are no methods that you can call on it which would contact the service. Would that not be a bit confusing?

Before our most recent changes that make functional objects extend infos we used to return `XXXInfo` objects while the methods were called simply `getXXX`. I think we should stick to this convention, adding the `Info` suffix does not make the method clearer to me.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java (hunk 20 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: Changed

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java (hunk 22 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: Reason is the same. The responses are identical up to the message and "location" attribute of "debugInfo". I switched it to location.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 19 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: Agreed. Refactored.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java (hunk 18 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DnsRpc.java]: Yes for maxResults. Not really for dnsName. The service allows two options: dnsName (e.g. for listing zones) and name (e.g. for listing dns records), so we need both options. I made a mistake of tossing both into dnsName. I am adding the option name option.

File: gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java (hunk 21 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java]: We did not have a special reason for it, probably just forgot. I will add when modifying for #610.

**Comment by mderka** [comment]
Re: Dns.getProjectInfo(), consistency sounds like a good enough reason. Since this touches layers above this rpc and some code that I have been working on in the mean time, I will handle it within a separate PR. Opening #619.

**Comment by mderka** [comment]
Ready for another check.

**Pull request merged**